grype/matcher/internal/result/results.go

```diff
@@ -213,6 +213,43 @@ func (s Set) Intersection(other Set) Set {
 	return out
 }
 
+// UpdateByIdentity applies an update function to results in the current set when they match
+// (by ID or alias) with entries in the incoming set. This is useful for replacing specific fields
+// (like fix information) while preserving other fields (like match details).
+//
+// For example, if the current set has CVE-2006-20001 and the incoming set has ALSA-2023:0852
+// with alias CVE-2006-20001, the updateFunc will be called to update the CVE entry with data
+// from the ALSA entry.
+func (s Set) UpdateByIdentity(incoming Set, updateFunc func(existing *Result, incoming Result)) Set {
```

Comment:
from our conversation, I feel like this (one day) could be a little more generic. This is a pretty specific functionality on a generic set implementation. We don't have to block on that now since this is in internal.

From what we chatted about the set intersectoin and getIdentity logic out and passed in as a "key function", thus this really changes from `UpdateByIdentity` to just `Update`. Odds are something like this can be done to `Merge`. TBD on removing of elements, but I feel that could be a separate call to `Filter` or `Remove` before hand.